File: Everything-app-backend/everything-app/src/main/java/com/Finn/everything_app/model/FinanceTransaction.java
package com.Finn.everything_app.model;

import jakarta.persistence.*;
import lombok.Data;

import java.time.LocalDate;
import java.time.LocalDateTime;


/**
 * Eine Buchung - von Hand eingetippt oder aus dem Bankimport.
 *
 * <p>Die Bank-Felder sind eine Erweiterung, kein Ersatz: manuelle Buchungen bleiben unveraendert
 * gueltig und tragen {@code externalId = null}, {@code bankAccount = null} und
 * {@link TransactionSource#MANUAL}.
 *
 * <p>{@link #type} bleibt ein {@code String} mit "EINNAHME"/"AUSGABE" - siehe
 * {@link TransactionType} fuer die Bruecke ins typsichere Enum.
 *
 * <p>Die hier deklarierten UNIQUE- und Index-Bedingungen wirken nur auf frisch erzeugten
 * Datenbanken (und in den H2-Tests). Auf einer gewachsenen Datenbank traegt ddl-auto=update sie
 * nicht mehr nach; dafuer gibt es db/manual/2026-08-06-finance-bank-import.sql.
 */
@Entity
@Table(name = "finance_transactions",
        uniqueConstraints = @UniqueConstraint(name = "uk_finance_tx_user_external",
                columnNames = {"user_id", "external_id"}),
        indexes = {
                @Index(name = "idx_finance_tx_user_date", columnList = "user_id, transaction_date"),
                @Index(name = "idx_finance_tx_bank_account", columnList = "bank_account_id"),
                @Index(name = "idx_finance_tx_contract", columnList = "contract_id")
        })
@Data
public class FinanceTransaction {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private Double amount;

    @Column(nullable = false, length = 50)
    private String type;

    @Column(nullable = false, length = 100)
    private String category;
    @Column(length = 100)
    private String subcategory;

    @Column(nullable = false, length = 500)
    private String description;

    @Column(name = "transaction_date", nullable = false)
    private LocalDate transactionDate;

    @Column(name = "payment_method", length = 50)
    private String paymentMethod;

    @Column(length = 500)
    private String tags;

    @Column(name = "receipt_url", length = 500)
    private String receiptUrl;

    @Column(name = "is_recurring")
    private Boolean isRecurring = false;

    @Column(name = "recurring_frequency", length = 50)
    private String recurringFrequency;

    // ==================== Bankimport ====================

    /**
     * Dedup-Schluessel des Imports; {@code null} bei manuellen Buchungen.
     *
     * <p>Enthaelt die {@code entry_reference} von Enable Banking - die PSD2 aber nur als optional
     * vorschreibt. Liefert eine Bank keine, muss der Sync einen berechneten Ersatz in dieselbe
     * Spalte schreiben (z.B. {@code "h:" + sha256(...)}, hex = 64 Zeichen); die Laenge 200 deckt
     * beide Faelle, und der eine UNIQUE-Index deckt beide Pfade.
     */
    @Column(name = "external_id", length = 200)
    private String externalId;

    /** Name der Gegenpartei (creditor/debtor) - Grundlage der Kategorisierung und Vertragserkennung. */
    @Column(length = 300)
    private String counterparty;

    /** Wertstellung. {@link #transactionDate} bleibt das Buchungsdatum. */
    @Column(name = "value_date")
    private LocalDate valueDate;

    /**
     * Herkunft der Buchung.
     *
     * <p>Trotz konzeptioneller Pflicht ohne {@code nullable = false}: ddl-auto=update setzt fuer
     * eine neue Spalte ein nacktes {@code ALTER TABLE ... ADD COLUMN ... NOT NULL} ohne Default ab.
     * Auf einer Tabelle mit Bestandszeilen scheitert das und bricht den gesamten
     * Schema-Update-Durchlauf ab - inklusive der neuen Tabellen. Lesende Stellen behandeln
     * {@code null} deshalb wie {@link TransactionSource#MANUAL}; die Handmigration zieht die
     * Altbestaende einmalig nach. Gleiches Muster wie bei {@link #isRecurring}.
     */
    @Enumerated(EnumType.STRING)
    @Column(length = 20)
    private TransactionSource source = TransactionSource.MANUAL;

    /**
     * Der Nutzer hat die Kategorie selbst gesetzt - die Auto-Kategorisierung fasst sie nicht mehr an.
     * Nullable aus demselben Grund wie {@link #source}; immer mit {@code Boolean.TRUE.equals(...)} lesen.
     */
    @Column(name = "category_locked")
    private Boolean categoryLocked = false;

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    // Relationships
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "budget_category_id")
    private BudgetCategory budgetCategory;

    /** Null bei manuellen Buchungen. Bewusst ohne cascade - vgl. Project.tasks. */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "bank_account_id")
    private BankAccount bankAccount;

    /** Von der Vertragserkennung gesetzt; parallel dazu wird {@link #isRecurring} auf true gezogen. */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "contract_id")
    private Contract contract;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: Everything-app-frontend/everything_app/lib/models/finance_transaction.dart
class FinanceTransaction {
  final int? id;
  final double amount;
  final String type; // 'EINNAHME' or 'AUSGABE'
  final String category;
  final String? subcategory;
  final String description;
  final DateTime transactionDate;
  final String? paymentMethod;
  final String? tags;
  final String? receiptUrl;
  final bool isRecurring;
  final String? recurringFrequency;
  final int? budgetCategoryId;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  // ── Bankimport ─────────────────────────────────────────────────────────────

  /// Name der Gegenpartei. Bei importierten Buchungen die große Zeile in der
  /// Liste, der Verwendungszweck die kleine. Bei getippten Buchungen `null`.
  final String? counterparty;

  /// 'MANUAL' oder 'BANK'. Bestandszeilen können `null` tragen und werden dann
  /// wie 'MANUAL' behandelt.
  final String? source;

  /// Der Nutzer hat die Kategorie selbst gesetzt - kein Vorschlag der Automatik.
  final bool categoryLocked;

  /// Gesetzt, wenn die Buchung zu einem erkannten Vertrag gehört.
  final int? contractId;

  final DateTime? valueDate;

  FinanceTransaction({
    this.id,
    required this.amount,
    required this.type,
    required this.category,
    this.subcategory,
    required this.description,
    required this.transactionDate,
    this.paymentMethod,
    this.tags,
    this.receiptUrl,
    this.isRecurring = false,
    this.recurringFrequency,
    this.budgetCategoryId,
    this.createdAt,
    this.updatedAt,
    this.counterparty,
    this.source,
    this.categoryLocked = false,
    this.contractId,
    this.valueDate,
  });

  bool get isIncome => type == 'EINNAHME';

  bool get isFromBank => source == 'BANK';

  /// Was in der Liste groß steht. Getippte Buchungen haben keine Gegenpartei,
  /// dort trägt die Beschreibung allein.
  String get title =>
      (counterparty != null && counterparty!.isNotEmpty) ? counterparty! : description;

  /// Die kleine Zeile darunter - leer, wenn sie nur den Titel wiederholen würde.
  String? get subtitle {
    if (counterparty == null || counterparty!.isEmpty) return null;
    if (description == counterparty) return null;
    return description;
  }

  FinanceTransaction copyWith({
    double? amount,
    String? type,
    String? category,
    String? subcategory,
    String? description,
    DateTime? transactionDate,
    String? paymentMethod,
    String? tags,
    bool? isRecurring,
    int? budgetCategoryId,
    String? counterparty,
    bool? categoryLocked,
  }) {
    return FinanceTransaction(
      id: id,
      amount: amount ?? this.amount,
      type: type ?? this.type,
      category: category ?? this.category,
      subcategory: subcategory ?? this.subcategory,
      description: description ?? this.description,
      transactionDate: transactionDate ?? this.transactionDate,
      paymentMethod: paymentMethod ?? this.paymentMethod,
      tags: tags ?? this.tags,
      receiptUrl: receiptUrl,
      isRecurring: isRecurring ?? this.isRecurring,
      recurringFrequency: recurringFrequency,
      budgetCategoryId: budgetCategoryId ?? this.budgetCategoryId,
      createdAt: createdAt,
      updatedAt: updatedAt,
      counterparty: counterparty ?? this.counterparty,
      source: source,
      categoryLocked: categoryLocked ?? this.categoryLocked,
      contractId: contractId,
      valueDate: valueDate,
    );
  }

  factory FinanceTransaction.fromJson(Map<String, dynamic> json) {
    return FinanceTransaction(
      id: json['id'],
      amount: (json['amount'] as num).toDouble(),
      type: json['type'],
      category: json['category'],
      subcategory: json['subcategory'],
      description: json['description'],
      transactionDate: DateTime.parse(json['transactionDate']),
      paymentMethod: json['paymentMethod'],
      tags: json['tags'],
      receiptUrl: json['receiptUrl'],
      isRecurring: json['isRecurring'] ?? false,
      recurringFrequency: json['recurringFrequency'],
      budgetCategoryId: json['budgetCategoryId'],
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
      counterparty: json['counterparty'],
      source: json['source'],
      categoryLocked: json['categoryLocked'] ?? false,
      contractId: json['contractId'],
      valueDate:
          json['valueDate'] != null ? DateTime.parse(json['valueDate']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'amount': amount,
      'type': type,
      'category': category,
      if (subcategory != null) 'subcategory': subcategory,
      'description': description,
      'transactionDate': transactionDate.toIso8601String().split('T')[0],
      if (paymentMethod != null) 'paymentMethod': paymentMethod,
      if (tags != null) 'tags': tags,
      if (receiptUrl != null) 'receiptUrl': receiptUrl,
      'isRecurring': isRecurring,
      if (recurringFrequency != null) 'recurringFrequency': recurringFrequency,
      if (budgetCategoryId != null) 'budgetCategoryId': budgetCategoryId,
      if (counterparty != null) 'counterparty': counterparty,
      // source, categoryLocked und contractId bleiben draußen: sie beschreiben
      // die Herkunft der Buchung und werden allein vom Backend gesetzt.
    };
  }
}
